server/sonar-server/src/main/java/org/sonar/server/computation/task/projectanalysis/issue/IssueCreationDateCalculator.java

```diff
@@ -0,0 +1,123 @@
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRule;
+import org.sonar.server.computation.task.projectanalysis.qualityprofile.ActiveRulesHolder;
+import org.sonar.server.computation.task.projectanalysis.scm.Changeset;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfo;
+import org.sonar.server.computation.task.projectanalysis.scm.ScmInfoRepository;
+import org.sonar.server.issue.IssueFieldsSetter;
+
+import static org.sonar.core.issue.IssueChangeContext.createScan;
+
+/**
+ * Calculates the creation date of an issue. Takes into account, that the issue
+ * might be raised by adding a rule to a quality profile.
+ */
+public class IssueCreationDateCalculator extends IssueVisitor {
+
+  private static final Logger LOGGER = Loggers.get(IssueCreationDateCalculator.class);
+
+  private final ScmInfoRepository scmInfoRepository;
+  private final IssueFieldsSetter issueUpdater;
+  private final AnalysisMetadataHolder analysisMetadataHolder;
+  private final IssueChangeContext changeContext;
+  private final ActiveRulesHolder activeRulesHolder;
+
+  public IssueCreationDateCalculator(AnalysisMetadataHolder analysisMetadataHolder, ScmInfoRepository scmInfoRepository,
+    IssueFieldsSetter issueUpdater, ActiveRulesHolder activeRulesHolder) {
+    this.scmInfoRepository = scmInfoRepository;
+    this.issueUpdater = issueUpdater;
+    this.analysisMetadataHolder = analysisMetadataHolder;
+    this.changeContext = createScan(new Date(analysisMetadataHolder.getAnalysisDate()));
+    this.activeRulesHolder = activeRulesHolder;
+  }
+
+  @Override
+  public void onIssue(Component component, DefaultIssue issue) {
+    if (issue.isNew() && ruleIsNew(issue)) {
+
+      // get scm information of file, if available
+      getChangeSet(component)
+        .map(scmInfo -> Optional
+
+          // if possible determine line change
+          .ofNullable(issue.getLine()).map(scmInfo::getChangesetForLine)
+
+          // otherwise determine file change
+          .orElseGet(scmInfo::getLatestChangeset))
+
+        // get the change's date
+        .map(Changeset::getDate)
+        .map(DateUtils::longToDate)
+
+        // use the change date as the creation date of the issue
+        .ifPresent(scmDate -> updateDate(issue, scmDate));
+    }
+  }
+
+  private boolean ruleIsNew(DefaultIssue issue) {
+    long ruleCreation = ruleCreation(issue);
+    Optional<Long> lastAnalysisOptional = lastAnalysis();
+
+    if (lastAnalysisOptional.isPresent()) {
+      return lastAnalysisOptional.get() < ruleCreation;
+    }
+
+    // special case: this is the first analysis of the project: use scm dates for all issues
+    return true;
+  }
+
+  private Optional<Long> lastAnalysis() {
+    return Optional.ofNullable(analysisMetadataHolder.getBaseAnalysis()).map(Analysis::getCreatedAt);
+  }
+
+  private long ruleCreation(DefaultIssue issue) {
+    ActiveRule activeRule = toJavaUtilOptional(activeRulesHolder.get(issue.getRuleKey()))
```

Comment:
we need to discuss whether this is more readable that the following (ping @simonbrandhof )
```
Optional<ActiveRule> activeRule = activeRulesHolder.get(issue.getRuleKey());
checkArgument(activeRule.isPresent(), "The rule %s raised an issue, but is not one of the active rules.", issue.getRuleKey().rule());
return activeRule.get();
```